rclcpp_lifecycle/src/state.cpp

```diff
@@ -14,49 +14,99 @@
 
 #include "rclcpp_lifecycle/state.hpp"
 
-#include <lifecycle_msgs/msg/state.hpp>
+#include <string>
 
-#include <rcutils/allocator.h>
-#include <rcutils/strdup.h>
-#include <rcl_lifecycle/data_types.h>
+#include "lifecycle_msgs/msg/state.hpp"
 
-#include <string>
+#include "rcutils/allocator.h"
+#include "rcutils/strdup.h"
+
+#include "rcl_lifecycle/data_types.h"
 
 namespace rclcpp_lifecycle
 {
 
-State::State()
-: State(lifecycle_msgs::msg::State::PRIMARY_STATE_UNKNOWN, "unknown")
+void
+copy_from(State & lhs, const State & rhs)
+{
+  if (lhs.owns_rcl_state_handle_) {
+    auto state_handle = reinterpret_cast<rcl_lifecycle_state_t *>(
+      rhs.allocator_.allocate(sizeof(rcl_lifecycle_state_t), rhs.allocator_.state));
+    state_handle->id = rhs.state_handle_->id;
+    state_handle->label = rcutils_strndup(
+      rhs.state_handle_->label,
+      strlen(rhs.state_handle_->label),
+      rhs.allocator_);
+    lhs.state_handle_ = state_handle;
+  } else {
+    lhs.state_handle_ = rhs.state_handle_;
```

Comment:
A previously allocated handle on `lhs` is being lost here.

Reply:
If we don't own the `state_handle_`, we can't simply deallocate the resource.

Reply:
But if we do we have to deallocate the resource before overwriting it.